### ISSUE TITLE
Istio Injection issue on vcluster >= 0.5.x

### ISSUE THREAD

**Issue by rmathagiarun** [loft-sh/vcluster#345]
After injecting istio into ingress-controller, the ingress pod doesn't seem to come up. This issue seems to occur specifically on vcluster >= 0.5. The same istio injection works fine on Vcluster 0.4.5.
 
K3s Version: v1.22.6-k3s1
Vcluster Version: 0.5.2 and 0.5.3
Istio Version: 1.11
Nginx Ingress Controller: 1.1.1
 
Sharing error logs for reference..
[ingress-controller-ingress.log](https://github.com/loft-sh/vcluster/files/7950265/ingress-controller-ingress.log)
[ingress-controller-istio-init.log](https://github.com/loft-sh/vcluster/files/7950266/ingress-controller-istio-init.log)
[ingress-controller-istio-proxy.log](https://github.com/loft-sh/vcluster/files/7950267/ingress-controller-istio-proxy.log)

**Comment by FabianKramm** [comment]
@rmathagiarun thanks for reporting this issue! We'll take a look at what the problem could be, would be great if you could also provide the vcluster log for containers vcluster and syncer

**Comment by rmathagiarun** [comment]
@FabianKramm 
Apologies for the delay as we had deleted the Vcluster before taking vcluster log backup. 
Sharing logs from another Vcluster which was used to reproduce the issue.
[vcluster2.log](https://github.com/loft-sh/vcluster/files/7992532/vcluster2.log)

**Comment by ramanNarasimhan77** [comment]
Hello @FabianKramm ,
I am a co-worker of @rmathagiarun  and I am also facing the same issue. 

I found this snippet in the syncer logs, which could probably help in diagnosing the issue
```[pod/vcluster2-0/syncer] E0203 06:21:36.849673       1 wrap.go:54] timeout or abort while handling: method=GET URI="/apis/admissionregistration.k8s.io/v1/validatingwebhookconfigurations?allowWatchBookmarks=true&labelSelector=istio.io%2Frev%3Ddefault&resourceVersion=1596435&timeout=7m5s&timeoutSeconds=425&watch=true" audit-ID="63d41fbd-57b2-4fba-b24c-6fddfb03ded3"
[pod/vcluster2-0/syncer] E0203 06:21:36.854285       1 webhook.go:155] Failed to make webhook authenticator request: tokenreviews.authentication.k8s.io is forbidden: User "system:serviceaccount:vcluster2:vc-vcluster2" cannot create resource "tokenreviews" in API group "authentication.k8s.io" at the cluster scope
[pod/vcluster2-0/syncer] E0203 06:21:36.854352       1 authentication.go:63] "Unable to authenticate the request" err="[[invalid bearer token, serviceaccounts \"istiod\" not found], tokenreviews.authentication.k8s.io is forbidden: User \"system:serviceaccount:vcluster2:vc-vcluster2\" cannot create resource \"tokenreviews\" in API group \"authentication.k8s.io\" at the cluster scope]"
[pod/vcluster2-0/syncer] E0203 06:21:36.854427       1 authentication.go:63] "Unable to authenticate the request" err="[[invalid bearer token, serviceaccounts \"istiod\" not found], tokenreviews.authentication.k8s.io is forbidden: User \"system:serviceaccount:vcluster2:vc-vcluster2\" cannot create resource \"tokenreviews\" in API group \"authentication.k8s.io\" at the cluster scope]"
[pod/vcluster2-0/syncer] E0203 06:21:36.854497       1 authentication.go:63] "Unable to authenticate the request" err="[[invalid bearer token, serviceaccounts \"istiod\" not found], tokenreviews.authentication.k8s.io is forbidden: User \"system:serviceaccount:vcluster2:vc-vcluster2\" cannot create resource \"tokenreviews\" in API group \"authentication.k8s.io\" at the cluster scope]"
```

Also there are lot of timeouts in the syncer logs
```
[pod/vcluster2-0/syncer] I0203 06:21:36.820967       1 logr.go:249] pod: ORG-mesh-controller-5dd85b45df-b5f76: delete physical pod vcluster2/ORG-mesh-controller-5dd85b45df-b5f76-x-APP-x-vcluster2, because virtual pod is being deleted
[pod/vcluster2-0/syncer] I0203 06:21:36.839559       1 logr.go:249] event-backward: create virtual event APP/ORG-mesh-controller-5dd85b45df-b5f76.16d0326a76fff835
[pod/vcluster2-0/syncer] E0203 06:21:36.847646       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/pods?allowWatchBookmarks=true&resourceVersion=1597052&timeout=9m21s&timeoutSeconds=561&watch=true" audit-ID="43e77230-0806-4be6-85ee-0b757f90f8b5"
[pod/vcluster2-0/syncer] E0203 06:21:36.847746       1 wrap.go:54] timeout or abort while handling: method=GET URI="/apis/apiextensions.k8s.io/v1/customresourcedefinitions?allowWatchBookmarks=true&resourceVersion=1547943&timeout=5m27s&timeoutSeconds=327&watch=true" audit-ID="2fc7ec55-9a22-42c2-b85e-01fd653d2c16"
[pod/vcluster2-0/syncer] E0203 06:21:36.847949       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/namespaces?allowWatchBookmarks=true&resourceVersion=1547943&timeout=5m7s&timeoutSeconds=307&watch=true" audit-ID="d5266813-6a3a-4cf0-bd74-a275ea6849b2"
[pod/vcluster2-0/syncer] E0203 06:21:36.848022       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/namespaces/APP/configmaps?allowWatchBookmarks=true&fieldSelector=metadata.name%3Distio-sidecar-injector&resourceVersion=1597006&timeout=8m30s&timeoutSeconds=510&watch=true" audit-ID="9a19db70-46f5-4e7a-9e29-627840854cd0"
[pod/vcluster2-0/syncer] E0203 06:21:36.848225       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/endpoints?allowWatchBookmarks=true&resourceVersion=1597054&timeout=5m25s&timeoutSeconds=325&watch=true" audit-ID="0671f445-3a46-4445-9cbb-565ffd8293e0"
[pod/vcluster2-0/syncer] E0203 06:21:36.848280       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/services?allowWatchBookmarks=true&resourceVersion=1597025&timeout=6m0s&timeoutSeconds=360&watch=true" audit-ID="3a5f4098-cc48-46a2-8338-f76177a7751c"
[pod/vcluster2-0/syncer] E0203 06:21:36.848325       1 wrap.go:54] timeout or abort while handling: method=GET URI="/apis/networking.k8s.io/v1/ingresses?allowWatchBookmarks=true&resourceVersion=1597010&timeout=8m1s&timeoutSeconds=481&watch=true" audit-ID="58dfbfec-182b-41fe-99af-b4baacc754a0"
[pod/vcluster2-0/syncer] E0203 06:21:36.848716       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/namespaces/APP/configmaps?allowWatchBookmarks=true&fieldSelector=metadata.name%3Distio&resourceVersion=1597006&timeout=5m57s&timeoutSeconds=357&watch=true" audit-ID="9074a6dd-42bf-4493-9797-8f9c01e3f76d"
[pod/vcluster2-0/syncer] E0203 06:21:36.848777       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/nodes?allowWatchBookmarks=true&resourceVersion=1596866&timeout=6m5s&timeoutSeconds=365&watch=true" audit-ID="a459c606-a09c-4d2a-a2cc-8b3bb7447640"
[pod/vcluster2-0/syncer] E0203 06:21:36.849041       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/secrets?allowWatchBookmarks=true&resourceVersion=1596140&timeout=9m6s&timeoutSeconds=546&watch=true" audit-ID="5ad9678c-de29-4c25-937a-e1f055a8adcd"
[pod/vcluster2-0/syncer] E0203 06:21:36.849241       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/configmaps?allowWatchBookmarks=true&resourceVersion=1597211&timeout=6m45s&timeoutSeconds=405&watch=true" audit-ID="67855cba-d6ce-4970-87ea-953235793142"
[pod/vcluster2-0/syncer] E0203 06:21:36.849455       1 wrap.go:54] timeout or abort while handling: method=GET URI="/apis/admissionregistration.k8s.io/v1/mutatingwebhookconfigurations?allowWatchBookmarks=true&fieldSelector=metadata.name%3Distio-sidecar-injector-APP&labelSelector=istio.io%2Frev%3Ddefault&resourceVersion=1596426&timeoutSeconds=559&watch=true" audit-ID="77217f47-1f89-46e3-b7a7-e9490949d3e6"
[pod/vcluster2-0/syncer] E0203 06:21:36.849517       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/namespaces/APP/secrets?allowWatchBookmarks=true&labelSelector=istio%2FmultiCluster%3Dtrue&resourceVersion=1596140&timeout=5m16s&timeoutSeconds=316&watch=true" audit-ID="accefabd-ba7a-45d9-96c1-20d63900e811"
[pod/vcluster2-0/syncer] E0203 06:21:36.849572       1 wrap.go:54] timeout or abort while handling: method=GET URI="/api/v1/namespaces/APP/secrets?allowWatchBookmarks=true&labelSelector=istio%2FmultiCluster%3Dtrue&resourceVersion=1596140&timeout=9m23s&timeoutSeconds=563&watch=true" audit-ID="5d6e42e5-f747-4c80-bb09-fa9b9e517d02"
[pod/vcluster2-0/syncer] E0203 06:21:36.849619       1 wrap.go:54] timeout or abort while handling: method=GET URI="/apis/networking.k8s.io/v1/ingressclasses?allowWatchBookmarks=true&resourceVersion=1547943&timeout=8m30s&timeoutSeconds=510&watch=true" audit-ID="28955040-6516-4a63-97ee-f742a7add71f"
```

Can you please provide some guidance for us to overcome this issue so that we can switch to newer vcluster versions?

**Comment by FabianKramm** [comment]
@ramanNarasimhan77 @rmathagiarun thanks for the additional info, seems like vcluster is trying to access the host cluster with a virtual cluster request. Are you overriding the `KUBE_CONFIG` environment variable or doing any other modifications to the vcluster deployment?

**Comment by rmathagiarun** [comment]
@FabianKramm  No, we follow the same process for both the versions. We export the Kubeconfig and use it for deploying our resources.

**Comment by FabianKramm** [comment]
@rmathagiarun I see, could you send us the mutated vcluster pod yaml (via `kubectl get pods -n MY_VCLUSTER_NAMESPACE -l app=vcluster -o yaml`)

**Comment by FabianKramm** [comment]
In addition would be also interesting if it works with the new `v0.6.0-beta.0` again

**Comment by rmathagiarun** [comment]
> @rmathagiarun I see, could you send us the mutated vcluster pod yaml (via `kubectl get pods -n MY_VCLUSTER_NAMESPACE -l app=vcluster -o yaml`)

[vcluster2_manifest.yaml.txt](https://github.com/loft-sh/vcluster/files/8022423/vcluster2_manifest.yaml.txt)

**Comment by FabianKramm** [comment]
@rmathagiarun seems like the pod is running fine and no istio container is injected there, is this the correct vcluster pod?

**Comment by ramanNarasimhan77** [comment]
Hello @FabianKramm ,
I see there is some disconnect here, I will try to make things clear.

Our workflow is 
a) create a vcluster, connect to it and get the kubeconfig
b) deploy our application into the vcluster

As part of our application we install Istio service mesh, and an Ingress controller which routes traffic into our application within the vcluster.
We inject Istio proxy side car into this Ingress controller.

The sidecar injection is happening fine on vcluster versions < 0.5.1 but on newer vcluster versions, the Ingress controller pod does not come up. 

@rmathagiarun  has shared the Ingress controller logs in the very first post

**Comment by FabianKramm** [comment]
@ramanNarasimhan77 okay thanks for the information. So just to be clear when you mean < 0.5.1, does that mean it works with 0.5.0 but not with 0.5.1? The ingress controller logs look fine to me, there is no error message in there, so I'm not exactly sure what to do here. Maybe it would be a good idea if you could give us a step by step guide how we can reproduce the problem.

**Comment by ramanNarasimhan77** [comment]
@FabianKramm  

I am sorry for my confusing comment and taking ample time to come up with a response. 

Following are the vcluster versions that we have tested our application against.

| Vcluster Version  | Test Status |
| ------------- | ------------- |
| 0.4.5  | Installation is successful  |
| 0.5.0  | Installation fails  |
| 0.5.1  | Installation fails  |
| 0.5.2  | Installation fails  |
| 0.6.0  | Installation fails  |

We use proprietary Istio helm chart in our organization, so I am not really sure how to give steps to reproduce this issue. I tried installing Open source Istio service Mesh into a namespace in a Vcluster and then installed a sample workload in another namespace and was able to get Istio injection working fine. 

However, since the installation works fine on `0.4.5` and does not work on any other subsequent `vcluster` version; I am not able to come to any working hypothesis on why this problem occurs in recent vcluster versions.

In our installation process we do the following steps
* Create a namespace on a Host cluster  and use it to provision a vcluster
* Within a namespace on the Vcluster we install our application which has a `nginx ingress controller` and `istio service mesh`.
* In the installation process we have a `Post install hook` which triggers a restart of the `nginx ingress controller` deployment. During this restart, istio-proxy side car container is supposed to be injected into the `nginx ingress controller pods`. This step is failing because the `istio-proxy` is not able to authenticate and connect to the `istio service mesh`

We have previously shared logs from the failed installation attempts in this same Issue. 

I understand that is really hard to diagnose this issue without a way to reproduce this, but would it be possible to take another look at this and give us some pointers on what has changed from `0.4.5` to `0.5.0` (release from which the installation starts to fail)

**Comment by rmathagiarun** [comment]
We were able to fix this issue by changing Istio configurations in our proprietary Istio helm chart. Thanks @FabianKramm for all the support. Closing this issue.

**Comment by FabianKramm** [comment]
@rmathagiarun great to hear this was resolved, I'm just curious, what was the solution to solving this problem as it might also benefit others?

**Comment by rmathagiarun** [comment]
@FabianKramm 

It was due to BoundServiceAccountTokenVolume enabled by default in k8s 1.21. When this is enabled, user needs to install istio by setting the global.jwtPolicy=third-party-jwt and then istio will configure new projected volume for separated SA Token (aud = istio-ca) in helm template, if we still keep to use "first-party-jwt" in istio, istio will report authN error log for BoundServiceAccountTokenVolume feature will set the "aud" to "https://kubenetes.default.svc.cluster.local/" which mean the target is k8sAPI server instead of istio.

Hence, we updated the global.jwtPolicy=third-party-jwt  to fix the issue.
